avocado/core/test.py

```diff
@@ -1093,13 +1094,39 @@ def fetch_asset(self, name, asset_hash=None, algorithm=None,
         :param locations: list of URLs from where the asset can be
                           fetched (optional)
         :param expire: time for the asset to expire
-        :raise EnvironmentError: When it fails to fetch the asset
-        :returns: asset file local path
+        :param find_only: When `True`, `fetch_asset` only looks for the asset
+                          in the cache, avoiding the download/move action.
+                          Defaults to `False`.
+        :param cancel_on_missing: whether the test should be canceled if the
+                                  asset was not found in the cache or if
+                                  `fetch` could not add the asset to the cache.
+                                  Defaults to `False`.
+        :raises OSError: when it fails to fetch the asset and
+                         `cancel_on_missing` is `False`.
+        :returns: asset file local path or `None` when asset was not found in
+                  the cache.
         """
         if expire is not None:
             expire = data_structures.time_to_seconds(str(expire))
-        return asset.Asset(name, asset_hash, algorithm, locations,
-                           self.cache_dirs, expire).fetch()
+
+        asset_obj = asset.Asset(name, asset_hash, algorithm, locations,
+                                self.cache_dirs, expire)
+
+        missing_asset_message = 'Missing asset %s' % name
+
+        if find_only:
+            asset_path = asset_obj.find_asset_file()
+            if asset_path is None and cancel_on_missing:
+                self.cancel(missing_asset_message)
+            return asset_path
+
+        try:
+            asset_path = asset_obj.fetch()
+        except OSError as e:
+            if cancel_on_missing:
+                self.cancel(missing_asset_message)
+            raise e
```

Comment:
This looks very good. Wondering if we shouldn't have the same effect for missing file on both `find_only=True` and `find_only=False` when `cancel_on_missing` is `False`. As of now, the former gives users a `None`, while the latter gives users an `OSError` exception. Wdyt?

Reply:
This is an interesting point. My only concern about this suggestion is that it would change the actual behavior of this method when `find_only=False` and `cancel_on_missing=False`. Users may, with the current upstream code, wait for an exception on their code when the `fetch_asset` method fails to add the asset to the cache. We always have the option to let the users know we will be changing this behavior in the upcoming release and go for the change. 

Another point is that as the method raises an exception coming from the Asset class when there is a problem fetching the asset, I'm afraid that, maybe, returning `None` masks the problem and confuses the users while debugging when `find_only=False` and `cancel_on_missing=False`.

Reply:
@apahim do you have any thoughts here? @clebergnu?

Reply:
I'd avoid breaking the API. As a user, with `cancel_on_missing=False`, I'd expect an `OSError` when the file is not found, no matter if `find_only` is `True` or `False`.

Reply:
Ah, I see. From your previous comment, I thought the suggestion was to return `None` in either case.

Okay, I don't have a broad opinion about returning `None` or throwing `OSError` when `find_only=True`. Thinking like a user, I would prefer to deal with `None` than handling an exception, but that is my option.

Reply:
My two cents here:

When you use `None` as return on fails, is harder to see why the method has failed. Because of that, I usually prefer exceptions.

Reply:
Okay, let's go for the exception in v4. Thanks, folks!